wiki/Client Security.md

```diff
@@ -0,0 +1,95 @@
+
+These two naive approaches are unsatisfactory because they are missing a few critical elements:
+
+1. They did not use randomness.
+2. Their position mapping function for each leaf is invertible.
+3. They did not expand the search space for brute-force attacker (confined within M<sup>2</sup>)
+
+To address these issues, we introduce the following technique:
+
+1. For each leaf, we use the corresponding OTP as the seed for a random number generator, to generate 4 consecutive numbers between [1, 4M]. Note this mapping is deterministic (hence repeatable) but is not invertible.
+2. We break down the leaf into 4 parts of 8-byte chunks, and store the chunks at the position given by the 4 random numbers generated in the previous step.
+3. If any two chunks fall into the same position, they are stored in a list at that position. For brevity we call this scenario a "collision".
+4. For each pair of neighbors, we swap the position of their chunks since we only need the neighbor to complete the Merkle Proof, not the leaf itself.
+
+At runtime, the client can easily recover the positions of each chunk by repeating (1). If there is a list consisting of more than one element at any position, the client can simply enumerate all possible combinations using each of the element in the list. In practice, the chance that any position has more than 1 or 2 elements in the list is very slim, since our the size of our random space in (1) equals to the number of elements we want to generate. We can further increase the space to increase sparsity and reduce collision. After recovering the chunks, the client can complete the Merkle Proof and proceed as usual. If there are multiple combinations of chunks due to collision, the client may test for each combination of whether it is correct, by computing the current leaf using **GenLeaf** and the user provided OTP, then check whether the hash of the concatenation of the current leaf and the neighbor chunk-combination candidate equals to the required hash value in the parent node corresponding to current timestamp or index.
+
+With this approach the number of operations required for simple brute-force is increased to (4M)<sup>8</sup> since the attacker would have to enumerate all possible combinations of chunks for each pair of neighbors and test whether they produce the required hash value in a any parent node. There may be more efficient brute-force attacks which we have not explored or analyzed.
```

Comment:
From @ivan-homoliak-sutd : 

> 4M does not seem correct. When you generate the 4 random numbers by PRNG, you can put the range of even 2^256 (by using hash) instead and store the elements in the mapping, and according to your formula, it'd mean unbelievable (2^256M)^8. However, the entropy of one OTP is only 2^20 ~ 2.5B, which, to me, is also manifested to PRNG you use. So, in fact, you have only 1M options => (1M)^8 ~ 160bits; which is pretty good.
> 
> Assuming that the left sibling was already submitted, this degrades to (1M)^4 ~ 80bits, which is still strong enough to be brute-forced in 30s or 60s (until the expiration of the right sibling).

Reply:
~4M comes from rough size of a dense array storing the chunks. For the array to be dense, its size should roughly be ~1M codes multiply by 4 chunks, i.e. 4M.  Even if we use a large range for the random number generator (e.g. 400M instead of 4M), it would be only as good as 4M because most positions would be zero. The result array would be sparse, and its equivalent dense representation would still only be roughly 4M long.   

Consider further, that if we generate more than 1M codes. Because there are only 1M OTPs in each 6-digit code, we would start to observe repetitions in chunks. The resulting dense array would still be effectively 4M in size for the reason that two or more repeating chunks are interchangeable, hence are only as good as one chunk and should be only counted as one entry in a dense array.

However, in a sparse representation where random range is set to a high value (again consider using 400M as an example), we may fill in the zero values with random noise. This construct may be effective and increase the number of operations required for simple brute-force to a very high value. More effective brute-force might exist, since we have not analyzed brute-force resistance of each component involved throughly (random number generator, whether patterns can be recovered from the scrambled layout, etc.). Note further that no matter how effective the scrambled technique is, we are limited to the 20-byte randomness in OTP Seed.